src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+                        found |= true;
+                    }
+                    // can't be both output (writeonly) and marked read only
+                    else if (hasReadOnlyArray && isOutputParam)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ArrayOutputParamMarkedRead, method.GetLocation(), method.Identifier, param.Identifier));
+                        found |= true;
+                    }
+                    // must have some indication of ReadOnly or WriteOnly
+                    else if (!hasWriteOnlyArray && !hasReadOnlyArray && !isOutputParam) 
+                    { 
+                        context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ArrayParamNotMarked, method.GetLocation(), method.Identifier, param.Identifier
+                            , hasWriteOnlyArray, hasReadOnlyArray, isOutputParam));
+                        found |= true;
+                    }
+                }
+                // Non-array types shouldn't have attributes meant for arrays
+                else if (hasWriteOnlyArray || hasReadOnlyArray)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.NonArrayMarked, method.GetLocation(), method.Identifier, param.Identifier));
+                    found |= true;
+                }
+            }
+
+            return found;
+        }
+
+#endregion
+
+        #region CheckingMethods
+
+        /// <summary>
+        /// Loops over each method declared in the given class and checks for various diagnostics
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="classDeclaration"></param>
+        /// <returns>True iff any of the diagnostics checked for are found.</returns>
+        public bool ClassHasInvalidMethods(ref GeneratorExecutionContext context, ClassDeclarationSyntax classDeclaration)
+        {
+            bool found = false;
+            IEnumerable<MethodDeclarationSyntax> methods = classDeclaration.ChildNodes().OfType<MethodDeclarationSyntax>();
+
+            Dictionary<string, bool> methodsHasAttributeMap = new Dictionary<string, bool>();
+            
+            /* we can't throw the diagnostic as soon as we see a second overload without an attribute, 
+             *   as there could be a third overload with the default attribute
+             * we use this map to keep track of these cases, and after we have checked all methods, 
+             *   we check the elements of the map to raise diagnostics for those that didn't get attributed */
+            Dictionary<string, Diagnostic> overloadsWithoutAttributeMap = new Dictionary<string, Diagnostic>();
+
+            foreach (MethodDeclarationSyntax method in methods.Where(MethodIsPublic))
+            {
+                found |= CheckParamsForArrayAttributes(method, ref context);

```

Comment:
I wonder if once one of these evaluates to true, whether the other functions get called or not?

Reply:
What do you mean?